### PR TITLE
JDK-8027711: Unify wildcarding syntax for CompileCommand and CompileOnly

### PR DESCRIPTION
**Pull request by tobiasholenstein**
At the moment `CompileCommand` and `CompileOnly` use different syntax for matching methods. 

### Old CompileOnly format
- matching a **method name** with **class name** and **package name**:
`-XX:CompileOnly=package/path/Class.method`
`-XX:CompileOnly=package/path/Class::method`
`-XX:CompileOnly=package.path.Class::method`
BUT NOT `-XX:CompileOnly=package.path.Class.method`

- just matching a **single method name**:
`-XX:CompileOnly=.hashCode`
`-XX:CompileOnly=::hashCode`
BUT NOT `-XX:CompileOnly=hashCode`

- Matching **all method names** in a **class name** with **package name**
`-XX:CompileOnly=java/lang/String`
BUT NOT `-XX:CompileOnly=java/lang/String.`
BUT NOT `-XX:CompileOnly=java.lang.String`
BUT NOT `-XX:CompileOnly=java.lang.String::` (This is actually a bug) 
BUT NOT `-XX:CompileOnly=String`
BUT NOT `-XX:CompileOnly=String.`
BUT NOT `-XX:CompileOnly=String::`

- Matching **all method names** in a **class name** with **NO package name**
`-XX:CompileOnly=String`
BUT NOT `-XX:CompileOnly=String.`
BUT NOT `-XX:CompileOnly=String::`

- There is a bug when `CompileOnly` ends with `::` where the `CompileOnly` is just ignored
e.g. `-XX:CompileOnly=String::` compiles as many methods as when omitting the `-XX:CompileOnly=` command

### CompileCommand=compileonly format
`CompileCommand` allows two different forms for paths: 
- `package/path/Class.method`
- `package.path.Class::method`

In contrary to `CompileOnly` `CompileCommand` supports wildcard matching using `*`. `*` can appear at the beginning and/or end of a `package.path.Class` and `method` name. 

 Valid forms: 
`-XX:CompileCommand=compileonly,*.lang.*::*shCo*`
`-XX:CompileCommand=compileonly,*/lang/*.*shCo*` 
`-XX:CompileCommand=compileonly,java.lang.String::*`
`-XX:CompileCommand=compileonly,*::hashCode`
`-XX:CompileCommand=compileonly,*ng.String::hashC*`
`-XX:CompileCommand=compileonly,*String::hash*`

 Invalid forms (Error: Embedded * not allowed):
 `-XX:CompileCommand=compileonly,java.*.String::has*Code`

### Use CompileCommand syntax for CompileOnly
At the moment, in some cases it is not possible to just take pattern used with `CompileOnly` and plug it into compile command file. Syntax used by CompileOnly is also not very intuitive. 

`CompileOnly` is convenient because it's shorter to write and takes lists of patterns, whereas `CompileCommand` only takes one pattern per command. 

With this PR `CompileOnly` becomes an alias for `CompileCommand=compileonly` with possibility to take lists as input. 

New syntax for `CompileOnly`:
- `-XX:CompileOnly=pattern1,pattern2`
is an **alias** for: 
- `-XX:CompileCommand=compileonly,pattern1 -XX:CompileCommand=compileonly,pattern2`

### Handling invalid syntax
Before `CompileOnly` just ignored invalid syntax. `CompileCommand` at least prints an error massage for invalid patterns:
```
CompileOnly: An error occurred during parsing
Error: Could not parse method pattern
Line: 'pattern'
```
Since `CompileOnly` now maps to `CompileCommand` is also prints the error message for invalid inputs. 
In the future `CompileCommand` (and `CompileOnly`) parsing errors could exit the VM https://bugs.openjdk.org/browse/JDK-8282797

### Changed test cases
In the following we mean with `-XX:CompileOnly=oldPattern` -> `-XX:CompileOnly=newPattern` that we changed the tests from the `oldPattern` to the `newPattern`:
- `Class.method` -> `Class::method` AND `package/path/Class.method` -> `package.path.Class::method`
Prefer the `package.path.Class::method` format because it is used by `-XX:+PrintCompilation`

- `Class` -> `Class::*`
The `CompileCommand` format requires the `::` to define if `Class` is a class name or a method name. 

- `::method` -> `*::method`
The `CompileCommand` format requires the `*`(wildcard) if no class name is given. 

- `package/path/Class` -> `package.path.Class::*`
The `CompileCommand` format requires the `*`(wildcard) if no method name is given. 
Prefer the `package.path.Class::method` format because it is used by `-XX:+PrintCompilation`

- `::get,::get1` -> `*Class::get*`
The `CompileCommand` format requires the `*`(wildcard) if no method name is given. Therefore `*::get` would work as well, but this matches many other methods as well like `java.util.HashMap::get`.
`*Class::get,*Class::get1` matches the wanted class name - `*Class::get*` is just the short form. 

- `package/path/Class::method` -> `package.path.Class::method`
The old format of `CompileOnly` combining `/` with `::`. 
The `CompileCommand` is either `package/path/Class.method` or `package.path.Class::method`. 

- _BUG_: `package.path.Class::` -> `package.path.Class::*`
There was a bug in the old format of `CompileOnly` : when the pattern ended with `::` the `CompileOnly` was just ignored and all methods compiled. The `CompileCommand` format requires the `*`(wildcard) if no method name is given.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change requires CSR request [JDK-8308287](https://bugs.openjdk.org/browse/JDK-8308287) to be approved
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issues
 * [JDK-8027711](https://bugs.openjdk.org/browse/JDK-8027711): Unify wildcarding syntax for CompileCommand and CompileOnly (**Enhancement** - P4) ⚠️ Issue is already resolved. Consider making this a "backport pull request" by setting the PR title to `Backport <hash>` with the hash of the original commit. See [Backports](https://wiki.openjdk.org/display/SKARA/Backports).
 * [JDK-8308287](https://bugs.openjdk.org/browse/JDK-8308287): Unify syntax of CompileOnly and CompileCommand (**CSR**)

### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to [d6fca2b8](https://git.openjdk.org/jdk/pull/13802/files/d6fca2b8266462673d68f7f1e13e726a4bb35b9c)
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**)
 * [Christian Hagedorn](https://openjdk.org/census#chagedorn) (@chhagedorn - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/13802/head:pull/13802` \
`$ git checkout pull/13802`

Update a local copy of the PR: \
`$ git checkout pull/13802` \
`$ git pull https://git.openjdk.org/jdk.git pull/13802/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 13802`

View PR using the GUI difftool: \
`$ git pr show -t 13802`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/13802.diff">https://git.openjdk.org/jdk/pull/13802.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/13802#issuecomment-1551533840)

### CODE REVIEW

**Review by vnkozlov — APPROVED** [review]
Thank you for fixing this finally!

FTR. We planned to do this for long time. Main motivations: unify syntax and catch invalid commands.

**Review by TobiHartmann — APPROVED** [review]
Thanks for taking care of this tedious changes, Toby. Looks good to me too!

File: test/hotspot/jtreg/compiler/loopopts/Test8211698.java [hunk 1]
```diff
@@ -54,4 +54,3 @@ public void test() {
         }
     }
 }
-
```
Comment (chhagedorn) [inline test/hotspot/jtreg/compiler/loopopts/Test8211698.java:57]: You can leave this empty line in at the end of the file.

**Review by chhagedorn — APPROVED** [review]
Nice unification! I just have some small comments. Otherwise, looks good!

File: test/hotspot/jtreg/compiler/loopopts/Test8211698.java (hunk 1 above)
Comment (tobiasholenstein) [reply test/hotspot/jtreg/compiler/loopopts/Test8211698.java]: done

**Review by chhagedorn — APPROVED** [review]
Update looks good!

**Comment by tobiasholenstein** [comment]
Thanks for the reviews @vnkozlov , @chhagedorn and @TobiHartmann

**Review by chhagedorn — APPROVED** [review]

**Review by chhagedorn — APPROVED** [review]

**Comment by vnkozlov** [comment]
GHA testing failed. Please look.
```
CompileOnly: An error occurred during parsing
Error: Could not parse method pattern
Line: 'TestZeroTripGuardShared'
```

**Review by chhagedorn — APPROVED** [review]
Update and testing results look good now!

**Review by TobiHartmann — APPROVED** [review]
Still looks good.

**Pull request closed without merging**